### PR TITLE
QARTOD Spike Test: alternative differential method 

### PR DESCRIPTION
**Pull request by JessyBarrette**
This PR present an implementation of the spike detection algorithm presented within the issue #48 .

This Pull Request essentially add an optional input method to the QARTOD spike test that and add an optional differential method for detecting spikes. If not specified, the standard method average method suggested by QARTOD is used.

A series of tests are also added to the test module to handle the different methods and the method input.

### CODE REVIEW

**Review by kwilcox — CHANGES_REQUESTED** [review]

File: ioos_qc/qartod.py [hunk 1]
```diff
@@ -480,17 +488,31 @@ def spike_test(inp : Sequence[N],
     original_shape = inp.shape
     inp = inp.flatten()
 
-    # Calculate the average of n-2 and n
-    ref = np.zeros(inp.size, dtype=np.float64)
-    ref[1:-1] = (inp[0:-2] + inp[2:]) / 2
-    ref = np.ma.masked_invalid(ref)
+    # Apply different method
+    if method is 'average':
```
[the remark this answers is not in the document]
Comment (JessyBarrette) [reply ioos_qc/qartod.py:492]: Oh yeah nice catch! Thanks easy fix

File: ioos_qc/qartod.py [hunk 2]
```diff
@@ -480,17 +488,31 @@ def spike_test(inp : Sequence[N],
...
+        # Calculate the average of n-2 and n
+        ref = np.zeros(inp.size, dtype=np.float64)
+        ref[1:-1] = (inp[0:-2] + inp[2:]) / 2
+        ref = np.ma.masked_invalid(ref)
+
+        # Calculate the (n-1 - ref) difference
+        diff = np.abs(inp - ref)
+    elif method is 'differential':
```
[the remark this answers is not in the document]
Comment (JessyBarrette) [reply ioos_qc/qartod.py:500]: Will do

File: ioos_qc/qartod.py [hunk 3]
```diff
@@ -480,17 +488,31 @@ def spike_test(inp : Sequence[N],
...
+
+        # Find the minimum variation prior and after the n value
+        diff = np.ma.zeros(inp.size, dtype=np.float64)
+        diff[1:-1] = np.minimum(np.abs(ref[:-1]), np.abs(ref[1:]))
+
+        # Make sure that only the record (n) where the difference prior and after are opposite are considered
+        with np.errstate(invalid='ignore'):
+            diff[1:-1][ref[:-1]*ref[1:] >= 0] = 0
+    else:
+        raise ValueError('"'+str(method)+'" method is unknown, "average" or "differential" methods are available')
```
[the remark this answers is not in the document]
Comment (JessyBarrette) [reply ioos_qc/qartod.py:511]: For sure!

**Comment by JessyBarrette** [comment]
I pushed the corrections! Thanks for your feedbacks @kwilcox

**Pull request merged**